internal/provisioner/kops_provisioner_cluster.go

```diff
@@ -443,40 +443,54 @@ func (provisioner *KopsProvisioner) DeleteCluster(cluster *model.Cluster, awsCli
 	}
 	defer kops.Close()
 
-	err = kops.UpdateCluster(kopsMetadata.Name, kops.GetOutputDirectory())
-	if err != nil {
-		return err
-	}
+	// Use these vars to keep track of which resources need to be deleted.
+	var skipDeleteKops, skipDeleteTerraform bool
 
-	terraformClient, err := terraform.New(kops.GetOutputDirectory(), provisioner.s3StateStore, logger)
-	if err != nil {
-		return err
-	}
-	defer terraformClient.Close()
+	logger.Info("Deleting cluster")
 
-	err = terraformClient.Init(kopsMetadata.Name)
+	_, err = kops.GetCluster(kopsMetadata.Name)
 	if err != nil {
-		return err
+		logger.WithError(err).Error("Failed kops get cluster check: proceeding assuming kops and terraform resources were never created")
+		skipDeleteKops = true
+		skipDeleteTerraform = true
 	}
 
-	err = verifyTerraformAndKopsMatch(kopsMetadata.Name, terraformClient, logger)
-	if err != nil {
-		// There are a few reasons why this may fail including an failed cluster
-		// creation. For now, let's log an error if there is one and proceed.
-		// TODO: make this better.
-		logger.WithError(err).Error("Proceeding with cluster deletion despite failing terraform output match check")
+	if !skipDeleteKops {
+		err = kops.UpdateCluster(kopsMetadata.Name, kops.GetOutputDirectory())
+		if err != nil {
+			return errors.Wrap(err, "failed to run kops update")
+		}
 	}
 
-	logger.Info("Deleting cluster")
+	if !skipDeleteTerraform {
+		terraformClient, err := terraform.New(kops.GetOutputDirectory(), provisioner.s3StateStore, logger)
+		if err != nil {
+			return errors.Wrap(err, "failed to create terraform wrapper")
+		}
+		defer terraformClient.Close()
 
-	err = terraformClient.Destroy()
-	if err != nil {
-		return err
+		err = terraformClient.Init(kopsMetadata.Name)
+		if err != nil {
+			return errors.Wrap(err, "failed to init terraform")
+		}
+
+		err = verifyTerraformAndKopsMatch(kopsMetadata.Name, terraformClient, logger)
+		if err != nil {
+			skipDeleteTerraform = true
```

Comment:
Good eye. It's not actually read after this assignment, but it was put there on purpose.

If we change something in the future where some other terraform tasks are performed much later in the delete process and we need to check whether to skip terraform or not, this will become relevant.

Reply:
Good enough for me. :+1: